[torch] redirect to bmm when tensor1 dim>=3 but not contiguous

jspark1105 opened the pull request:
Summary:
matmul for [B, M, K] x [K, N] was mapped to mm by folding the first 2dim of tensor1 to [BxM, K] x [K, N] but when M and K are transposed it's better to use BMM to avoid data movement.

We could generalize the condition we don't fold (see more details in the comment) but being conservative here to be cautious about potential unintended regression.

Test Plan:
In the following simple test case, before this diff

0.00652953577041626 0.003044447898864746
Permutation takes about same time as GEMM

After this diff
0.002983328104019165 0.0030336639881134034
Permutation overhead essentially went away.

```
B = 128
M = 1024
N = 128
K = 1024

X = torch.rand(B, K, M).cuda()
b = torch.rand(N).cuda()
W = torch.rand(N, K).cuda()
X = X.permute(0, 2, 1)
Y = F.linear(X, W, b)

X_contiguous = X.contiguous()
Y_ref = F.linear(X_contiguous, W, b)

torch.testing.assert_close(Y, Y_ref)

t1, _ = benchmark_torch_function(F.linear, X, W, b, 0)

t2, _ = benchmark_torch_function(F.linear, X_contiguous, W, b, 0)

print(t1, t2)
```

Differential Revision: D34350990

lezcano reviewed (approved):
Good point!

A similar thing can be done in the matrix x batch case, which currently always copies, as it calls contiguous after transposing a contiguous matrix. cc @ngimel

ngimel commented:
Test failure seems real, probably because we are running on older cards. OpInfo dtype tests should be adjusted/skipped for bfloat16.

ngimel commented:
Unfortunately, even this limited heuristics isn't always beneficial https://github.com/pytorch/pytorch/issues/56891#issuecomment-830925396, additional factors like the sizes of the matrix to be flattened and the second arg have to be taken into account.

jspark1105 commented:
Thanks for comments! Will take a look at the test failures and potential regression cases.

ngimel commented:
We still can land even though small matrices regress I think, if the test errors are fixed. The error is likely due to spotty bfloat16 support on an older GPU, so some targeted test skips should be enough.

IvanYashchuk commented:
`test_sparse_addmm_cpu_bfloat16` fail on master because of this PR: https://github.com/pytorch/pytorch/pull/72846

jspark1105 committed:
[torch] do not fold bmm into mm when tensor1 dim==3 but not contiguous (#73115)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/73115

matmul for [B, M, K] x [K, N] was mapped to mm by folding the first 2dim of tensor1 to [BxM, K] x [K, N] but when M and K are transposed it's better to use BMM to avoid data movement.

We could generalize the condition we don't fold (see more details in the comment) but being conservative here to be cautious about potential unintended regression.

Test Plan:
In the following simple test case, before this diff

0.00652953577041626 0.003044447898864746
Permutation takes about same time as GEMM

After this diff
0.002983328104019165 0.0030336639881134034
Permutation overhead essentially went away.

```
B = 128
M = 1024
N = 128
K = 1024

X = torch.rand(B, K, M).cuda()
b = torch.rand(N).cuda()
W = torch.rand(N, K).cuda()
X = X.permute(0, 2, 1)
Y = F.linear(X, W, b)

X_contiguous = X.contiguous()
Y_ref = F.linear(X_contiguous, W, b)

torch.testing.assert_close(Y, Y_ref)

t1, _ = benchmark_torch_function(F.linear, X, W, b, 0)

t2, _ = benchmark_torch_function(F.linear, X_contiguous, W, b, 0)

print(t1, t2)
```

Differential Revision: D34350990

fbshipit-source-id: f3dc761e3766e0f6f78b61eddc6d3a38f1d8a6d7

jspark1105 commented:
After rebase and skipping bfloat16 for < sm53 tests are passing.
